CoreFeatures/HttpEndpoint/DeserializeStrategyPostMultipartFormData/src/main/java/info/smart_tools/smartactors/http_endpoint/deserialize_strategy_post_multipart_form_data/FileSavingStrategy.java

```diff
@@ -0,0 +1,161 @@
+                        throw new RuntimeException("Issues with IO has occurred.", e);
+                    }
+                }
+        );
+        this.onFileExistsActions.put(
+                1, (file, source) -> {
+                    throw new NotImplementedException();
+                }
+        );
+        this.onFileExistsActions.put(
+                2, (file, source) -> {
+                    try {
+                        file = validateFileExists(file, 0);
+                        saveFile(file, source);
+                    } catch (IOException e) {
+                        throw new RuntimeException("Issues with IO has occurred.", e);
+                    }
+                }
+        );
+        this.onFileExistsActions.put(
+                3, (file, source) -> {
+                    if (file.exists()) {
+                        throw new RuntimeException("File already exists.");
+                    } else {
+                        try {
+                            saveFile(file, source);
+                        } catch (IOException e) {
+                            throw new RuntimeException("Issues with IO has occurred.", e);
+                        }
+                    }
+                }
+        );
+    }
+
+    @Override
+    @SuppressWarnings("unchecked")
+    public <T> T resolve(Object... args) throws StrategyException {
+        FileUpload file = (FileUpload) args[0];
+        Integer onFileExists = null;
+        String targetDirectory = null;
+        if (args.length > 1) {
+            Integer value = (Integer) args[1];
+            onFileExists = null == value || value > 3 || value < 0 ? 0 : value;
+        }
+        if (args.length > 2) {
+            targetDirectory = (String) args[2];
+        }
+        IActionTwoArgs<File, FileUpload> action = this.onFileExistsActions.get(onFileExists);
+        File newFile = new File(targetDirectory + File.separator + file.getName());
+        try {
+            validatePathAccessibility(targetDirectory, newFile.getParent());
+            action.execute(newFile, file);
+        } catch (InvalidArgumentException | ActionExecutionException | IOException e) {
+            throw new StrategyException(e);
+        }
+
+        return (T) newFile;
+    }
+
+    private File validateFileExists(final File file, final int count) throws IOException {
+        String path = file.getParentFile().getCanonicalPath();
+        File newFile = count != 0 ?
+                new File(
+                        path + File.separator + getFileName(file) +
+                        LEFT_BRACKET + count + RIGHT_BRACKET +
+                        (!getFileExtension(file).isEmpty() ? EXTENSION_DELIMITER + getFileExtension(file) : "")
+                ) :
+                new File(file.getCanonicalPath());
+        if (!newFile.exists()) {
+            return newFile;
+        } else {
+            int newCount = count + 1;
+            return validateFileExists(file, newCount);
```

Comment:
Probably, a bad point for recursion. Here it can be easy replaced with a loop.

Reply:
fixed